### ISSUE TITLE
Compilation errors on unreachable match patterns

### ISSUE THREAD

**Issue by hasithaa** [ballerina-platform/ballerina-spec#724]
In the implementation, we do data flow analysis and detects if there are patterns unreachable and gives compilation errors, i.e : 

```
public function main() returns error? {
    
    int|string x = "";
    match x {
        var [a] => {} // Compile time error: pattern will not be matched
        _ => {}
        var a => { } // Compile time error: unreachable pattern
    }
} 
```

We have a similar situation with conditions, i.e

```
if true  { }  // Works, But Should we give an error here?

if 5 == 5  // Works, But Should we give an error here?

if 5 is int // Compile time error: unnecessary condition: expression will always evaluate to 'true'
}

```

These kinds of extra validations are not clearly mentioned in the spec. Having these validations may create problems in the future.

**Comment by jclark** [comment]
I think the spec and compiler should be in sync: the compiler shouldn't reject code that the spec allows.

I find it useful to use `if` with things that are known true/false to get the same effect as commenting code out or back in again, particularly since we don't have multi-line comments.

If you have `match E { }` where E has type T, and you have a match pattern P and there is no value of type T that matches P, then I think that should be an error.

For unreachable patterns, I would like to make that work coherently with type narrowing. In your example
1. `_` matches all values of type `int|string`
2. the narrowed type for `x` following that pattern should be `never`
3. there is no value belonging to `never` that matches `var a` (or indeed any pattern)

This is complicated by `if` in `match` clauses.

Related to this is the question of  exhaustiveness (whether the match patterns exhaustively deal with all the possibilities being matches). For example, this should be OK:

```
enum Color { R, G, B };

function colorName(Color c) returns string {
   match c {
      R => { return "red"; }
      G => { return "green"; }
      B => { return "blue"; }
    }
   // this is unreachable so no possibility of falling off
}
```

But if I left out on the cases, it should get a compile error.

**Comment by jclark** [comment]
This is addressed by #913 and #887
